File: app/api/DBConnector.php
<?php
/**
 * Created by PhpStorm.
 * User: Gene
 * Date: 12/11/2016
 * Time: 6:36 AM
 */

/**
 * Class DBConnector
 * Wrapper for the PHP mysqli database connector object
 * Main query methods are:
 *    getResultArray(): returns <array(<array(key=>value)>)>
 *    getResultRow(): returns <array(key=>value)>
 *    query(): returns <int|bool>
 */
class DBConnector {
   private $conn;
   private $host;
   private $user;
   private $password;
   private $database;

   /**
    * DBConnector constructor.
    * May be called with no parameters. In this case the connect() method called at the end of the constructor will
    * do nothing and we will need to establish a db connection later on by manually calling connect() with the correct
    * database parameters
    * @param $host
    * @param $user
    * @param $password
    * @param $database
    */
   function __construct($host=null, $user=null, $password=null, $database=null) {
      $this->conn = null;
      $this->host = $host;
      $this->user = $user;
      $this->password = $password;
      $this->database = $database;

      $this->connect();
   }

   /**
    * Initiate MySQL database connection.
    * Method may be called with no parameters. In this case it will use the parameters previously set in the constructor
    * @param $host
    * @param $user
    * @param $password
    * @param $database
    */
   public function connect($host=null, $user=null, $password=null, $database=null) {
      // if new parameters were specified, set the class variables to them
      if ($host && $user && $password && $database) {
         $this->host = $host;
         $this->user = $user;
         $this->password = $password;
         $this->database = $database;
      }

      // try to initiate a sql connection if db parameters are set
      if ($this->host && $this->user && $this->password && $this->database) {

         $this->conn = mysqli_connect($this->host, $this->user, $this->password);

         if (!mysqli_select_db($this->conn, $this->database)) {
            print "Not connected to database.";
            $this->conn = null;
         }
      }
   }

   /**
    * Check to make sure $this->conn is not null and if it is, attempt to open it
    */
   private function checkConnection() {
      if ($this->conn != null) {
         return true;
      } else {
         $this->connect();
         if ($this->conn == null) {
            return false;
         } else {
            return true;
         }
      }
   }

   /**
    * Get a list of results from the database, multiple rows
    * @param $query
    * @return array
    */
   public function getResultArray($query) {
      $result_rows = array();
      
      if (! $this->checkConnection()) {
         return $result_rows;
      }
      
      $result = mysqli_query($this->conn, $query);

      while ($result_row = mysqli_fetch_assoc($result)) {
         $result_rows[] = $result_row;
      }
      return $result_rows;
   }

   /**
    * Get a single row result from the database
    * @param $query
    * @return array|null
    */
   public function getResultRow($query) {
      $result_row = array();

      if (! $this->checkConnection()) {
         return $result_row;
      }
      
      $result = mysqli_query($this->conn, $query);

      $result_row = mysqli_fetch_assoc($result);

      return $result_row;
   }

   /**
    * Execute a query that does not return any data such as an INSERT or UPDATE statement
    * returns true if the query succeeded and false if something went wrong
    * @param $query
    * @return bool|mysqli_result|void
    */
   public function query($query) {
      if (! $this->checkConnection()) {
         return false;
      }
      $result = mysqli_query($this->conn, $query);

      return $result;
   }

   /**
    * Returns the last insert id from the mysqli connection object
    * @return int|string
    */
   public function lastInsertId() {
      return mysqli_insert_id($this->conn);
   }

   /**
    * User the mysqli object to escape a string and returns the result
    * @param $string: String to be escaped
    * @return string: Escaped string
    */
   public function escape($string) {
      return mysqli_real_escape_string($this->conn, $string);
   }
}

File: data_loader/DBConnector.py
import mysql.connector
from mysql.connector import errorcode
import MySQLdb


class DBConnector:
    """
    A basic wrapper for the python sql connector class
    Has a couple different ways to set connection params and some connection checking before running queries
    Basic functionality:
        query():          run a general query
        getResultRow():   get one row from the db in the form of a dictionary
        getResultArray(): get a list of dictionaries back, with a dictionary representig each row
    """
    conn = None
    host = None
    user = None
    password = None
    database = None

    def __init__(self, host=None, user=None, password=None, database=None):
        """
        Constructor
        May be called with no parameters. In this case the connect() method called at the end of the constructor will
        do nothing and we will need to establish a db connection later on by manually calling connect() with the correct
        database parameters
        :param host:
        :param user:
        :param password:
        :param database:
        """
        self.host = host
        self.user = user
        self.password = password
        self.database = database

        self.connect()

    def connect(self, host=None, user=None, password=None, database=None):
        """
        Initiate MySQL database connection.
        Method may be called with no params. In this case it will use the parameters previously set in the constructor
        :param host:
        :param user:
        :param password:
        :param database:
        :return:
        """
        # if new parameters were specified, set the class variables to them
        if host and user and password and database:
            self.host = host
            self.user = user
            self.password = password
            self.database = database

        if self.host and self.user and self.password and self.database:

            try:                                    # try to connect to the database
                self.conn = mysql.connector.connect(user=self.user,
                                                    password=self.password,
                                                    host=self.host,
                                                    database=self.database)
            except mysql.connector.Error as error:  # handle connection and wrong db errors

                if error.errno == errorcode.ER_ACCESS_DEMIED_ERROR:
                    print 'Something went wrong with your access permissions. Please check your username and password'
                elif error.errno == errorcode.ER_BAD_DB_ERROR:
                    print("Database does not exist")
                else:
                    print(error)

    def disconnect(self):
        """
        Close the MySQL connection
        """
        self.conn.close

    def checkConnection(self):
        """
        Some redundant checking to make sure the SQL connection is open before making any requests
        If the connection is not open, attempt to open it
        :return:
        """
        if self.conn is not None:
            return True
        else:
            self.connect()          # attempt to popen a new sql connection if none is found
            if self.conn is None:   # unable to open a sql connection
                return False
            else:                   # successfully opened a new sql connection
                return True

    def getResultArray(self, query, params=None):
        """
        Get a list of results from the database, multiple rows
        :param query: Query String
        :param params: Optional parameters to run a parametrized query

        :return: 2 dimensional list
        """
        result_array = []

        if not self.checkConnection():
            return result_array

        cursor = self.conn.cursor(dictionary=True)
        if params is None:
            cursor.execute(query)
        else:
            cursor.execute(query, params)

        for row_dict in cursor:
            result_array.append(row_dict)

        cursor.close()

        return result_array

    def getResultRow(self, query, params=None):
        """
        Get a single row result from the database
        :param query: Query String
        :param params: Optional parameters to run a parametrized query
        :return: list
        """
        result_row = []

        if not self.checkConnection():
            return result_row

        cursor = self.conn.cursor(dictionary=True)
        if params is None:
            cursor.execute(query)
        else:
            cursor.execute(query, params)

        result_row = cursor[0]

        cursor.close()

        return result_row

    def query(self, query, params=None):
        """
        Execute a query that does not return any data such as an INSERT or UPDATE statement
        returns true if the query succeeded and false if something went wrong
        :param query: Query String
        :param params: Optional parameters to run a parametrized query
        :return: boolean or int - last insert id
        """
        result = False

        if not self.checkConnection():
            return result

        cursor = self.conn.cursor()
        try:
            if params is None:
                cursor.execute(query)
            else:
                cursor.execute(query, params)

            self.conn.commit()
        except mysql.connector.Error as error:
            print error.msg
            result = False
        else:
            if cursor.lastrowid is not None:
                result = cursor.lastrowid
            else:
                result = True

        cursor.close()

        return result

    def queryMany(self, query, param_list):
        """
        Execute many parametrized queries at once
        returns true if the query succeeded and false if something went wrong
        :param param_list: list of tuples to plug into each query
        :param query:
        :return: boolean or int - last insert id
        """
        result = False

        if not self.checkConnection():
            return result

        cursor = self.conn.cursor()
        try:
            cursor.executemany(query, param_list)
            self.conn.commit()
        except mysql.connector.Error as error:
            print error.msg
            result = False
        else:
            if cursor.lastrowid is not None:
                result = cursor.lastrowid
            else:
                result = True

        cursor.close()

        return result

    def escape(self, string):
        """
        Use mysql_real_escape to escape the input string and return the result
        Escaping is handled automatically by the python sql connector. This method is here just in case
        :param string: Input string
        :return: Escaped result
        """
        escaped_string = MySQLdb.escape_string(string)

        return escaped_string
